README.md
# [simter-http](https://github.com/simter/simter-http)

## 参考

- [OkHttp](http://square.github.io/okhttp/)
- [RxJava](https://github.com/ReactiveX/RxJava)


CHANGELOG.md
# simter-http changelog

## 3.0.0 - 2022-06-21

- Upgrade to simter-dependencies-3.0.0 (jdk-17)

## 2.0.0 - 2020-11-19

- Upgrade to simter-dependencies-2.0.0

## 1.1.0 - 2019-07-02

No code changed, just polishing maven config and unit test.

- Use JUnit5|AssertJ instead of JUnit4|Hamcrest
- Change parent to simter-dependencies-1.2.0

## 1.0.0 - 2019-01-08

- Just align version

## 0.5.0 - 2018-01-05

- Just align version

## 0.4.0 - 2018-01-05

- Just centralize-version

## 0.3.0 - 2017-12-13

- initial

src/test/java/tech/simter/http/OkHttpUtilsTest.java
package tech.simter.http;

import okhttp3.OkHttpClient;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * @author RJ
 */
class OkHttpUtilsTest {
  @Test
  void timeout() {
    OkHttpClient client = OkHttpUtils.newClient();
    assertEquals(20000, client.connectTimeoutMillis());
    assertEquals(60000, client.readTimeoutMillis());
    assertEquals(20000, client.writeTimeoutMillis());
  }
}

src/test/java/tech/simter/http/UserAgentUtilsTest.java
package tech.simter.http;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * @author RJ
 */
class UserAgentUtilsTest {
  @Test
  void defaultUA() {
    assertEquals(UserAgentUtils.get("default"), UserAgentUtils.get());
    assertEquals(UserAgentUtils.get("not-exists"), UserAgentUtils.get());
    assertEquals(UserAgentUtils.get("xp-ie6"), UserAgentUtils.get("xp-ie"));
  }

  @Test
  void win10UA() {
    assertEquals(UserAgentUtils.get("win10-ie11"), UserAgentUtils.get("win10-ie"));
    assertThat(UserAgentUtils.get("win10-ie11")).contains("Windows NT 10.0", "rv:11.0");

    assertEquals(UserAgentUtils.get("win10-chrome46"), UserAgentUtils.get("win10-chrome"));
    assertThat(UserAgentUtils.get("win10-chrome46")).contains("Windows NT 10.0", "Chrome/46");

    assertEquals(UserAgentUtils.get("win10-firefox52"), UserAgentUtils.get("win10-firefox"));
    assertThat(UserAgentUtils.get("win10-firefox52")).contains("Windows NT 10.0", "Firefox/52");
  }

  @Test
  void win7UA() {
    assertEquals(UserAgentUtils.get("win7-ie9"), UserAgentUtils.get("win7-ie"));
    assertThat(UserAgentUtils.get("win7-ie9")).contains("Windows NT 6.1", "MSIE 9.0");

    assertEquals(UserAgentUtils.get("win7-chrome46"), UserAgentUtils.get("win7-chrome"));
    assertThat(UserAgentUtils.get("win7-chrome46")).contains("Windows NT 6.1", "Chrome/46");

    assertEquals(UserAgentUtils.get("win7-firefox45"), UserAgentUtils.get("win7-firefox"));
    assertThat(UserAgentUtils.get("win7-firefox45")).contains("Windows NT 6.1", "Firefox/45");
  }

  @Test
  void xpUA() {
    assertEquals(UserAgentUtils.get("xp-chrome46"), UserAgentUtils.get("xp-chrome"));
    assertThat(UserAgentUtils.get("xp-chrome46")).contains("Windows NT 5.1", "Chrome/46");

    assertEquals(UserAgentUtils.get("xp-ie6"), UserAgentUtils.get("xp-ie"));
    assertThat(UserAgentUtils.get("xp-ie6")).contains("Windows NT 5.1", "MSIE 6.0");

    assertEquals(UserAgentUtils.get("xp-firefox52"), UserAgentUtils.get("xp-firefox"));
    assertThat(UserAgentUtils.get("xp-firefox52")).contains("Windows NT 5.1", "Firefox/52");
  }

  @Test
  void androidUA() {
    assertEquals(UserAgentUtils.get("android-chrome44"), UserAgentUtils.get("android-chrome"));
    assertThat(UserAgentUtils.get("android-chrome44")).contains("Android", "Chrome/44");
  }

  @Test
  void iphoneUA() {
    assertEquals(UserAgentUtils.get("iphone-safari10"), UserAgentUtils.get("iphone-safari"));
    assertThat(UserAgentUtils.get("iphone-safari10")).contains("iPhone OS", "Version/10");
  }
}

src/main/java/tech/simter/http/OkHttpUtils.java
package tech.simter.http;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.InetSocketAddress;
import java.net.Proxy;
import java.util.concurrent.TimeUnit;

/**
 * @author RJ 2017-03-29
 */
public final class OkHttpUtils {
  private final static Logger logger = LoggerFactory.getLogger(OkHttpUtils.class);
  public static final String SIMTER_PROXY_ENABLE = "simter_proxy_enable";
  public static final String SIMTER_PROXY_HOSTNAME = "simter_proxy_hostname";
  public static final String SIMTER_PROXY_PORT = "simter_proxy_port";
  private static Proxy proxy;

  /**
   * Set a global proxy
   *
   * @param hostname The Host name
   * @param port     The port number
   */
  public static void proxy(String hostname, int port) {
    proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(hostname, port));
  }

  /**
   * get a proxy through system environment config
   * <ul>
   * <li>'simter.proxy.enable' = 'true' : return a new proxy through 'simter.proxy.hostname' and 'simter.proxy.port' config, default to 127.0.0.1:8888</li>
   * <li>'simter.proxy.enable' = 'false' or no 'simter.proxy.enable' config : no proxy use, return null</li>
   * </ul>
   *
   * @return the proxy
   */
  public static Proxy getEnvProxy() {
    // Whether to enable the proxy
    boolean enableProxy;
    try {
      enableProxy = Boolean.parseBoolean(System.getenv(SIMTER_PROXY_ENABLE));
    } catch (Exception e) {
      logger.warn(e.getMessage(), e);
      enableProxy = false;
    }
    if (!enableProxy) return null; // don't use proxy

    // get hostname, default to "127.0.0.1"
    String hostname = null;
    try {
      hostname = System.getenv(SIMTER_PROXY_HOSTNAME);
    } catch (Exception e) {
      logger.warn(e.getMessage(), e);
    }
    if (hostname == null) hostname = "127.0.0.1";

    // get port, default to 8888
    String port_ = null;
    try {
      port_ = System.getenv(SIMTER_PROXY_PORT);
    } catch (Exception e) {
      logger.warn(e.getMessage(), e);
    }
    int port;
    try {
      port = port_ == null ? 8888 : Integer.parseInt(port_);
    } catch (Exception e) {
      logger.warn(e.getMessage(), e);
      port = 8888;
    }

    logger.info("use proxy {}:{}", hostname, port);
    return new Proxy(Proxy.Type.HTTP, new InetSocketAddress(hostname, port));
  }

  /**
   * Create a new OkHttpClient instance with simter default config:
   * <ul>
   * <li>1) proxy config through system environment</li>
   * <li>2) connectTimeout = 20s, readTimeout = 30s, writeTimeout = 20s</li>
   * </ul>
   *
   * @return the client instance
   */
  public static OkHttpClient newClient() {
    return newClientBuilder().build();
  }


  /**
   * Create a new OkHttpClient.Builder instance with simter default config:
   * <ul>
   * <li>1) proxy config through system environment</li>
   * <li>2) connectTimeout = 20s, readTimeout = 60s, writeTimeout = 20s</li>
   * </ul>
   *
   * @return the builder instance
   */
  public static OkHttpClient.Builder newClientBuilder() {
    OkHttpClient.Builder builder = new OkHttpClient.Builder();

    // proxy
    if (proxy != null) {    // use custom proxy
      builder.proxy(proxy);
    } else {                // use default proxy from system environment config
      Proxy envProxy = getEnvProxy();
      if (envProxy != null) builder.proxy(envProxy);
    }

    // timeout: OkHttp default is connectTimeout = 10s, readTimeout = 10s, writeTimeout = 10s, pingInterval = 0;
    builder.connectTimeout(20, TimeUnit.SECONDS)
      .readTimeout(60, TimeUnit.SECONDS)
      .writeTimeout(20, TimeUnit.SECONDS);

    return builder;
  }

  /**
   * Create a new Request.Builder instance with simter default config:
   * <ul>
   * <li>Header: User-Agent = ${UserAgentUtils.get()}</li>
   * </ul>
   *
   * @return the builder instance
   */
  public static Request.Builder newRequestBuilder() {
    return new Request.Builder().addHeader("User-Agent", UserAgentUtils.get());
  }
}

src/main/java/tech/simter/http/UserAgentUtils.java
package tech.simter.http;

import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Browser UserAgent collection and tools
 * <p>see http://rongjih.blog.163.com/blog/static/335744612011821111612884</p>
 *
 * @author RJ 2017-03-29
 */
public class UserAgentUtils {
  private static Map<String, String> agents = new ConcurrentHashMap<>();

  static {
    // initial default user-agents

    // win10
    agents.put("win10-ie11", "Mozilla/5.0 (Windows NT 10.0; WOW64; Trident/7.0; rv:11.0) like Gecko");
    agents.put("win10-ie", agents.get("win10-ie11"));

    agents.put("win10-edge", "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) " +
      "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.79 Safari/537.36 Edge/14.14393");

    agents.put("win10-chrome46",
      "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.86 Safari/537.36");
    agents.put("win10-chrome56",
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.87 Safari/537.36");
    agents.put("win10-chrome", agents.get("win10-chrome46"));

    agents.put("win10-firefox52", "Mozilla/5.0 (Windows NT 10.0; WOW64; rv:52.0) Gecko/20100101 Firefox/52.0");
    agents.put("win10-firefox", agents.get("win10-firefox52"));

    // win7
    agents.put("win7-ie8", "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.1; WOW64; Trident/4.0)");
    agents.put("win7-ie9", "Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Win64; x64; Trident/5.0)");
    agents.put("win7-ie10", "Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.1; WOW64; Trident/6.0)");
    agents.put("win7-ie", agents.get("win7-ie9"));

    agents.put("win7-chrome46",
      "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.71 Safari/537.36");
    agents.put("win7-chrome", agents.get("win7-chrome46"));

    agents.put("win7-firefox45", "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:45.0) Gecko/20100101 Firefox/45.0");
    agents.put("win7-firefox", agents.get("win7-firefox45"));


    // winXP
    agents.put("xp-ie8", "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.1; WOW64; Trident/4.0)");
    agents.put("xp-ie7", "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1)");
    agents.put("xp-ie6", "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1)");
    agents.put("xp-ie", agents.get("xp-ie6"));

    agents.put("xp-chrome46",
      "Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.71 Safari/537.36");
    agents.put("xp-chrome", agents.get("xp-chrome46"));

    agents.put("xp-firefox52", "Mozilla/5.0 (Windows NT 5.1; rv:52.0) Gecko/20100101 Firefox/52.0");
    agents.put("xp-firefox", agents.get("xp-firefox52"));

    // iphone
    agents.put("iphone-safari10", "Mozilla/5.0 (iPhone; CPU iPhone OS 10_2_1 like Mac OS X) AppleWebKit/602.4.6 " +
      "(KHTML, like Gecko) Version/10.0 Mobile/14D27 Safari/602.1");
    agents.put("iphone-safari", agents.get("iphone-safari10"));

    // android
    agents.put("android-chrome44", "Mozilla/5.0 (Linux; Android 6.0.1; SAMSUNG SM-N9200 Build/MMB29K) " +
      "AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/4.2 Chrome/44.0.2403.133 Mobile Safari/537.36");
    agents.put("android-chrome", agents.get("android-chrome44"));

    // default
    agents.put("default", agents.get("win7-ie"));
  }

  /**
   * Get a UserAgent by key
   * <p>
   * If key not exists, return a default UserAgent
   *
   * @param key the key
   * @return the UserAgent string
   */
  public static String get(String key) {
    return agents.containsKey(key) ? agents.get(key) : agents.get("default");
  }

  /**
   * Get a Default UserAgent
   * <p>
   * equals to get("default")
   *
   * @return the default UserAgent string
   */
  public static String get() {
    return agents.get("default");
  }

  /**
   * All available keys
   *
   * @return the UserAgent keys
   */
  public static Set<String> keys() {
    return agents.keySet();
  }

  /**
   * Add a UserAgent
   *
   * @param key       the key
   * @param userAgent the userAgent
   */
  public static void add(String key, String userAgent) {
    agents.put(key, userAgent);
  }
}